src/commands/server/login.ts

```diff
@@ -0,0 +1,152 @@
+ * which is available at https://www.eclipse.org/legal/epl-2.0/
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ **********************************************************************/
+
+import { Command, flags } from '@oclif/command'
+import { string } from '@oclif/parser/lib/flags'
+import { cli } from 'cli-ux'
+import * as execa from 'execa'
+
+import { CheHelper } from '../../api/che'
+import { CheApiClient } from '../../api/che-api-client'
+import { CheServerLoginManager, LoginRecord } from '../../api/che-login-manager'
+import { KubeHelper } from '../../api/kube'
+import { cheNamespace, CHE_API_ENDPOINT_KEY, username, USERNAME_KEY } from '../../common-flags'
+import { OPENSHIFT_CLI } from '../../util'
+
+const REFRESH_TOKEN_KEY = 'refresh-token'
+const PASSWORD_KEY = 'password'
+export default class Login extends Command {
+  static description = 'log in to Eclipse Che server'
+
+  static args = [
+    {
+      name: CHE_API_ENDPOINT_KEY,
+      description: 'Eclipse Che server API endpoint',
+      env: 'CHE_API_ENDPOINT',
+      required: false // In case of login via oc token with admin rights
+    }
+  ]
+  static flags: flags.Input<any> = {
+    help: flags.help({ char: 'h' }),
+    chenamespace: cheNamespace,
+    [REFRESH_TOKEN_KEY]: string({
+      char: 't',
+      description: 'Keycloak refresh token',
+      env: 'CHE_KEYCLOAK_REFRESH_TOKEN',
+      required: false,
+    }),
+    [USERNAME_KEY]: username,
+    [PASSWORD_KEY]: string({
+      char: 'p',
+      description: 'Eclipse Che user password',
+      env: 'CHE_USER_PASSWORD',
+      required: false,
+    }),
+  }
+
+  static examples = [
+    'server:login https://che-che.apps-crc.testing/api -u <username> -p <password>',
+    'server:login che-che.apps-crc.testing -u <username>',
+    'server:login che.openshift.io -t <token>',
+  ]
+
+  async run() {
+    const { args, flags } = this.parse(Login)
+
+    const loginManager = await CheServerLoginManager.getInstance(this.config.configDir)
+
+    let cheApiClient: CheApiClient
+    let cheApiEndpoint: string | undefined = args[CHE_API_ENDPOINT_KEY]
+    if (!cheApiEndpoint) {
+      const kube = new KubeHelper(flags)
+      if (!await kube.hasReadPermissionsForNamespace(flags.chenamespace)) {
+        throw new Error('Please provide server API URL argument')
+      }
+      // Retrieve API URL from routes
+      const cheHelper = new CheHelper(flags)
+      cheApiEndpoint = await cheHelper.cheURL(flags.chenamespace) + '/api'
+      cli.info(`Using ${cheApiEndpoint} server API URL to log in`)
+      cheApiClient = CheApiClient.getInstance(cheApiEndpoint)
+    } else {
+      cheApiEndpoint = CheApiClient.normalizeCheApiEndpointUrl(cheApiEndpoint)
+      cheApiClient = CheApiClient.getInstance(cheApiEndpoint)
+      try {
+        await cheApiClient.checkCheApiEndpointUrl()
+      } catch (error) {
+        // Wrong API URL, try to guess, maybe base url is provided
+        if (!cheApiEndpoint.endsWith('api')) {
```

Comment:
I think we shouldn't do excess request, we can check `api` suffix before first request and avoid guessing.

Reply:
Or maybe `normalizeCheApiEndpointUrl` should check and apply `/api` if needed

Reply:
Unfortunately we cannot be sure that Che api will always be on `ahost/api`. It is in almost all cases, but we cannot predict all possible network configurations, so I would go first with what  user exactly provided and only if that failed, then try to guess. Another point for such behaviour is that login process is not time critical, so, I think, we can do such guessing, which is really helpful.

Reply:
> Or maybe normalizeCheApiEndpointUrl should check and apply /api if needed

No, because the task of normalization is to provide correctly formatted url and not guessing something (and not make any requests). Otherwise it should be called like `normilizeGuessQueryCheApi` which is weird.